core/block_validator.go

```diff
@@ -90,7 +90,7 @@ func (v *BlockValidator) ValidateState(block *types.Block, statedb *state.StateD
 		return fmt.Errorf("invalid bloom (remote: %x  local: %x)", header.Bloom, rbloom)
 	}
 	// Tre receipt Trie's root (R = (Tr [[H1, R1], ... [Hn, R1]]))
-	receiptSha := types.DeriveSha(receipts, new(trie.Trie))
+	receiptSha := types.DeriveSha(receipts, trie.NewStackTrie(nil))
```

Comment:
I fail to see how this can work. Don't you need to do something like this: 
```diff
 func DeriveSha(list DerivableList) common.Hash {
        keybuf := new(bytes.Buffer)
-       trie := new(trie.Trie)
-       for i := 0; i < list.Len(); i++ {
+       //trie := new(trie.ReStackTrie)
+       trie :=trie.NewReStackTrie()
+       for i := 1; i < list.Len() && i <= 0x7f; i++ {
                keybuf.Reset()
                rlp.Encode(keybuf, uint(i))
-               trie.Update(keybuf.Bytes(), list.GetRlp(i))
+               trie.TryUpdate(keybuf.Bytes(), list.GetRlp(i))
+       }
+       for i := 0; i < list.Len() && i < 1; i++ {
+               keybuf.Reset()
+               rlp.Encode(keybuf, uint(i))
+               trie.TryUpdate(keybuf.Bytes(), list.GetRlp(i))
+       }
+       for i := 0x80; i < list.Len(); i++ {
+               keybuf.Reset()
+               rlp.Encode(keybuf, uint(i))
+               trie.TryUpdate(keybuf.Bytes(), list.GetRlp(i))
        }
        return trie.Hash()
 }
```
?

Reply:
And I'm also wondering how this isn't failing any tests?